rfcs/0000-type-stability.md

```diff
@@ -0,0 +1,500 @@
+
+```ts
+const myPromise: Promise<{}> = dontCarePromise();
+```
+
+…then it broke on TS 3.5, with the compiler reporting an error ([playground][3.5-breakage-plaground]):
+
+> Type 'Promise<unknown>' is not assignable to type 'Promise<{}>'.
+>   Type 'unknown' is not assignable to type '{}'.
+
+This change could be mitigated by supplying a default type argument equal to the original value ([playground][3.5-mitigation-playground]):
+
+```ts
+function dontCarePromise(): Promise<{}> {
+  return new Promise((resolve) => {
+    resolve();
+  });
+}
+```
+
+This is a totally-backwards compatible bugfix-style change, and should be released in a bugfix/point release. Users can then just upgrade to the bugfix release *before* upgrading their own TypeScript version—and will experience *zero* impact from the breaking TypeScript change.
+
+Later, the default type argument `Promise<{}>` could be dropped and defaulted to the new value for a major release of the library when desired (per the policy [outlined below](#policy-for-supported-type-script-versions), giving it the new semantics. (Also see [<b>Opt-in future types</b>](#opt-in-future-types) below for a means to allow users to *opt in* to these changes before the major version.)
+
+[3.3-pre-breakage-playground]: https://www.typescriptlang.org/play/?ts=3.3.3&ssl=1&ssc=27&pln=1&pc=40#code/GYVwdgxgLglg9mABAEwVAwgQwE4FMAK2cAtjAM64AUAlIgN4CwAUIonlCNkmLgO6KES5KpTxk4AGwBuuWgF4AfPWatWYyTJoBuFYgC+1HUz3NmEBGSiJiAT0GkKALgFEHuADx09SuSjRY8e2FtIA
+[3.5-breakage-plaground]: https://www.typescriptlang.org/play/?ts=3.5.1&ssl=1&ssc=27&pln=1&pc=40#code/GYVwdgxgLglg9mABAEwVAwgQwE4FMAK2cAtjAM64AUAlIgN4CwAUIonlCNkmLgO6KES5KpTxk4AGwBuuWgF4AfPWatWYyTJoBuFYgC+1HUz3NmEBGSiJiAT0GkKALgFEHuADx09SuSjRY8e2FtIA
+[3.5-mitigation-playground]: https://www.typescriptlang.org/play/?ts=3.5.1#code/GYVwdgxgLglg9mABAEwVAwgQwE4FMAK2cAtjAM64AUAlAFyKEnm4A8A3gL4B8ibAsAChEiPFBDYkYXAHcGRUhUqU8ZOABsAbrmqIAvD35DhI3Ks1VqAbkHCOVwR0GCICMlETEAnowW56P5nZuPRQ0LDwAxSsgA
+
+##### “Downleveling” types
+
+When a new version of TypeScript results in backwards-incompatible *emit* to to the type definitions, as they did in [3.7][3.7-emit-change], the strategy of changing the types directly will not work. However, it is still possible to provide backwards-compatible types, using the combination of [downlevel-dts] and [typesVersions]. (In some cases, this may also require some manual tweaking of types, but this should be rare for most addons.)
+
+- [downlevel-dts] allows you to take a `.d.ts` file which is not valid for an earlier version of TypeScript (e.g. the changes to class field emit mentioned in [<b>Breaking Changes to Type Definitions</b>](#breaking-changes-to-type-definitions)), and emit a version which *is* compatible with that version. Specifically: it currently generates types compatible with TypeScript 3.4.[[see note 2](#notes)]
```

Comment:
Do we see evidence that this will be maintained and supported for all breakage that we may need it for in the future? How do the compatibility guarantees around this tool's output align with the guarantees we're looking to provide for Ember+TS users?

Reply:
Good callouts on maintenance/support for downlevel-dts. I intentionally bypassed those questions by revising to make *all* the mechanics for mitigations non-normative. We can't force people to use any of these anyway! We may choose at some point in the future to adopt a tool like downlevel-dts as a recommended and perhaps even *supported* tool for this, but we don't need to do so today.